aten/src/ATen/native/cpu/AdaptiveMaxPoolKernel.cpp

```diff
@@ -0,0 +1,341 @@
+    const Tensor& grad_input_,
+    const Tensor& grad_output_,
+    const Tensor& indices_) {
+  TORCH_CHECK(grad_output_.ndimension() == 4,
+              "adaptive max pooling backward with channels last format supports tensors with 4 dims.");
+  auto memory_format = at::MemoryFormat::ChannelsLast;
+  auto grad_input = grad_input_.contiguous(memory_format);
+  auto grad_output = grad_output_.contiguous(memory_format);
+  auto indices = indices_.contiguous(memory_format);
+
+  auto grad_input_data = grad_input.data_ptr<scalar_t>();
+  auto grad_output_data = grad_output.data_ptr<scalar_t>();
+  auto indices_data = indices.data_ptr<int64_t>();
+
+  int64_t nbatch = grad_input.size(0);
+  int64_t channels = grad_input.size(1);
+  int64_t input_height = grad_input.size(2);
+  int64_t input_width = grad_input.size(3);
+  int64_t output_height = grad_output.size(2);
+  int64_t output_width = grad_output.size(3);
+
+  // parallel on dim N
+  at::parallel_for(0, nbatch, 0, [&](int64_t begin, int64_t end) {
+    for (int64_t n = begin; n < end; n++) {
+      scalar_t* grad_input_ptr = grad_input_data + n * input_height * input_width * channels;
+      scalar_t* grad_output_ptr = grad_output_data + n * output_height * output_width * channels;
+      int64_t* indices_ptr = indices_data + n * output_height * output_width * channels;
+
+      for (int64_t oh = 0; oh < output_height; oh++) {
+        for (int64_t ow = 0; ow < output_width; ow++) {
+          scalar_t* gout = grad_output_ptr + oh * output_width * channels + ow * channels;
+          int64_t* ind = indices_ptr + oh * output_width * channels + ow * channels;
+          // TODO: gcc vectorization
+          for (int64_t c = 0; c < channels; c++) {
+            int64_t maxindex = ind[c];
+            grad_input_ptr[maxindex * channels + c] += gout[c];
+          }
+        }
+      }
+    }
+  });
+
+  if (!grad_input_.is_contiguous(memory_format)) {
+    grad_input_.copy_(grad_input);
+  }
+}
+
+void adaptive_max_pool2d_kernel_impl(
+    const Tensor& output,
+    const Tensor& indices,
+    const Tensor& input,
+    IntArrayRef output_size) {
+  switch (input.suggest_memory_format()) {
+    case at::MemoryFormat::Contiguous: {
+      AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "adaptive_max_pool2d", [&] {
+        cpu_adaptive_max_pool<scalar_t>(output, indices, input, output_size);
+      });
+      break;
+    }
+    case at::MemoryFormat::ChannelsLast: {
+      AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "adaptive_max_pool2d_channels_last", [&]{
+        cpu_adaptive_max_pool_channels_last<scalar_t>(output, indices, input, output_size);
+      });
+      break;
+    }
+    default:
+      TORCH_CHECK(false, "Unsupported memory format. Supports only ChannelsLast, Contiguous");
+  }
+}
+
+void adapative_max_pool2d_backward_kernel_impl(
+    const Tensor& grad_input,
+    const Tensor& grad_output,
+    const Tensor& indices) {
+  switch (grad_output.suggest_memory_format()) {
```

Comment:
this won't work properly, in many cases adaptiveMaxPooling has an output size of (1,1), and suggested memory format won't be ChannelsLast. You'd need to pass `input` as argument here, to get suggested memory format from `input`.

Reply:
right! I updated with "grad_input.suggest_memory_format()"

grad_input is created via:
```C++
set_output(0, input.sizes(), input.options().memory_format(input.suggest_memory_format()));
```
has the same memory format with input.